kanidmd/src/lib/idm/authsession.rs

```diff
@@ -1950,4 +1987,200 @@ mod tests {
+
+        let backup_codes = BackupCodes::new(code_set);
+
+        // add totp and backup codes also
+        let p = CryptoPolicy::minimum();
+        let cred = Credential::new_password_only(&p, pw_good)
+            .unwrap()
+            .update_totp(totp)
+            .update_backup_code(backup_codes)
+            .unwrap();
+
+        account.primary = Some(cred);
+
+        let (async_tx, mut async_rx) = unbounded();
+
+        // now check
+        // == two step checks
+
+        // Sending a PW first is an immediate fail.
+        {
+            let (mut session, _, pw_badlist_cache) =
+                start_password_mfa_session!(&mut audit, account, &webauthn);
+
+            match session.validate_creds(
+                &mut audit,
+                &AuthCredential::Password(pw_bad.to_string()),
+                &ts,
+                &async_tx,
+                &webauthn,
+                Some(&pw_badlist_cache),
+                &hs512,
+            ) {
+                Ok(AuthState::Denied(msg)) => assert!(msg == BAD_AUTH_TYPE_MSG),
+                _ => panic!(),
+            };
+        }
+        // check send wrong backup code, should fail immediate
+        {
+            let (mut session, _, pw_badlist_cache) =
+                start_password_mfa_session!(&mut audit, account, &webauthn);
+
+            match session.validate_creds(
+                &mut audit,
+                &AuthCredential::BackupCode(backup_code_bad),
+                &ts,
+                &async_tx,
+                &webauthn,
+                Some(&pw_badlist_cache),
+                &hs512,
+            ) {
+                Ok(AuthState::Denied(msg)) => assert!(msg == BAD_BACKUPCODE_MSG),
+                _ => panic!(),
+            };
+        }
+        // check send good backup code, should continue
+        //      then bad pw, fail pw
+        {
+            let (mut session, _, pw_badlist_cache) =
+                start_password_mfa_session!(&mut audit, account, &webauthn);
+
+            match session.validate_creds(
+                &mut audit,
+                &AuthCredential::BackupCode(backup_code_good.clone()),
+                &ts,
+                &async_tx,
+                &webauthn,
+                Some(&pw_badlist_cache),
+                &hs512,
+            ) {
+                Ok(AuthState::Continue(cont)) => assert!(cont == vec![AuthAllowed::Password]),
+                _ => panic!(),
+            };
+            match session.validate_creds(
+                &mut audit,
+                &AuthCredential::Password(pw_bad.to_string()),
+                &ts,
+                &async_tx,
+                &webauthn,
+                Some(&pw_badlist_cache),
+                &hs512,
+            ) {
+                Ok(AuthState::Denied(msg)) => assert!(msg == BAD_PASSWORD_MSG),
+                _ => panic!(),
+            };
+        }
+        // Can't process BackupCodeRemoval without the server instance
```

Comment:
That's fine, you just have to deque here and check that the removal message *was* sent.